Add MFA message to TDP

ibeckermayer opened the pull request:
Adds a TDP message of the form

**MFA JSON**
`| message type (10) | mfa_type byte | message_length uint32 | json []byte`

to hold mfa challenge and response json's (challenge when being sent from server --> client, response when sent the opposite way).

`mfa_type` is one of `'n'` for webauthn or `'u'` for u2f, in imitation of the existing, similar mechanism we have for distinguishing between the two in the corresponding proto messages.

## TODO

- [x]  The TDP rfd (0037) should be updated with a better written version of the above
- [x] [`DecodeMFAJson`](https://github.com/gravitational/teleport/blob/af86260316058525c928a3dbb76dbf8845a3056b/lib/srv/desktop/tdp/proto.go#L523-L551) (called from `lib/web/desktop.go::mfaDecode`) and [`lib/web/terminal.go::mfaDecode`](https://github.com/gravitational/teleport/blob/af86260316058525c928a3dbb76dbf8845a3056b/lib/web/terminal.go#L337-L372) both have very similar code blocks, whose logic might be extracted into a function somewhere sensible.

Commit:
adds an MfaJson message type to TDP and refactors promptMFAChallenge to be able to accept functions for encode and decode so that it can be used for both protobuf and tdp/json messages

Commit:
Merge branch 'smallinsky/mfa-for-rds-v2' into isaiah/tdp-mfa

zmb3 commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -443,6 +456,104 @@ func decodeClipboardData(in peekReader, maxLen uint32) (ClipboardData, error) {
+
+type MFAJson struct {
+	// MfaType should be one of WebsocketU2FChallengeByte or WebsocketWebauthnChallengeByte.
+	MfaType byte
```
```suggestion
	MFAType byte
```

zmb3 commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -47,6 +52,12 @@ const (
 	TypeClientUsername   = MessageType(7)
 	TypeMouseWheel       = MessageType(8)
 	TypeError            = MessageType(9)
+	TypeMFAJson          = MessageType(10)
```
```suggestion
	TypeMFAJSON         = MessageType(10)
```

Convention in Go is for acronyms to by all caps.

zmb3 commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -47,6 +52,12 @@ const (
+)
+
+const (
+	WebsocketU2FChallengeByte      = 'u'
```
```suggestion
	WebsocketU2FChallenge      = 'u'
```

zmb3 commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -443,6 +456,104 @@ func decodeClipboardData(in peekReader, maxLen uint32) (ClipboardData, error) {
+	return buf.Bytes(), nil
+}
+
+func DecodeMFAJson(in peekReader) (*MFAJson, error) {
```
```suggestion
func DecodeMFAJSON(in peekReader) (*MFAJson, error) {
```

zmb3 commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -443,6 +456,104 @@ func decodeClipboardData(in peekReader, maxLen uint32) (ClipboardData, error) {
+			},
+		}
+
+	default:
```
How about `case WEbsocketU2FChallengeByte` instead of using default?

You already verified that it's one of these options anyway.

zmb3 commented on lib/web/terminal.go:
```diff
@@ -339,7 +396,12 @@ func (t *TerminalHandler) issueSessionMFACerts(tc *client.TeleportClient, ws *we
 }
 
-func promptMFAChallenge(ws *websocket.Conn, wsLock *sync.Mutex) client.PromptMFAChallengeHandler {
+func promptMFAChallenge(
```
Please add godoc explaining what encode and decode are supposed to do.

zmb3 reviewed (commented):
Minor naming suggestions.

I'll look at your TODO next.

smallinsky commented on lib/web/terminal.go:
```diff
@@ -317,6 +317,63 @@ func (t *TerminalHandler) issueSessionMFACerts(tc *client.TeleportClient, ws *we
+
+	mfaDecode := func(bytes []byte, envelopeType string) (*authproto.MFAAuthenticateResponse, error) {
+		envelope := &Envelope{}
+		if err = proto.Unmarshal(bytes, envelope); err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		var mfaResponse *authproto.MFAAuthenticateResponse
+
+		// Convert from JSON to proto.
+		switch envelopeType {
+		case defaults.WebsocketWebauthnChallenge:
+			var webauthnResponse wanlib.CredentialAssertionResponse
+			if err := json.Unmarshal([]byte(envelope.Payload), &webauthnResponse); err != nil {
+				return nil, trace.Wrap(err)
+			}
+			mfaResponse = &authproto.MFAAuthenticateResponse{
+				Response: &authproto.MFAAuthenticateResponse_Webauthn{
+					Webauthn: wanlib.CredentialAssertionResponseToProto(&webauthnResponse),
+				},
+			}
+
+		default:
+			var u2fResponse u2f.AuthenticateChallengeResponse
+			if err := json.Unmarshal([]byte(envelope.Payload), &u2fResponse); err != nil {
+				return nil, trace.Wrap(err)
+			}
+			mfaResponse = &authproto.MFAAuthenticateResponse{
+				Response: &authproto.MFAAuthenticateResponse_U2F{
+					U2F: &authproto.U2FResponse{
+						KeyHandle:  u2fResponse.KeyHandle,
+						ClientData: u2fResponse.ClientData,
+						Signature:  u2fResponse.SignatureData,
+					},
+				},
+			}
+		}
+
+		return mfaResponse, nil
+	}
```
nit:
What would you say about defining ecplicty MFAChalangeEncoder and MFAChallangeDecore interfaces instaed of using inline function and having the  `MFAChalangeEncDec` and MFAChallnageRDSEncDec types:

```go
   type MFAChalangeEncDec interface {
      Encode(chal *auth.MFAAuthenticateChallenge, envelopeType string) ([]byte, error)
      Decode(bytes []byte, envelopeType string) (*authproto.MFAAuthenticateResponse, error)
   }
```
and typen just defining the Proto RDS MFA types like:
```go
type MFAChallnageProtoEncDec struct {
}

func (m *MFAChallnageProtoDecoder) Encode(chal *auth.MFAAuthenticateChallenge, envelopeType string) ([]byte, error) {
  ...
}

func (m *MFAChallnageProtoDecoder)  Decode(bytes []byte, envelopeType string) (*authproto.MFAAuthenticateResponse, error)([]byte, error) {
  ...
}
```
```go
type MFAChallnageRDSEncDec struct {
}

func (m *MFAChallnageRDSEncDec) Encode(chal *auth.MFAAuthenticateChallenge, envelopeType string) ([]byte, error) {
  ...
}

func (m *MFAChallnageRDSEncDec)  Decode(bytes []byte, envelopeType string) (*authproto.MFAAuthenticateResponse, error)([]byte, error) {
  ...
}
```
and during the `promptMFAChallenge` call we can just pass a proper type:
`promptMFAChallenge(ws, t.wsLock, &MFAChallnageRDSEncDec{})`

zmb3 replied:
That's exactly what I was thinking. I'm going to put this change together now and update the PR.

ibeckermayer replied:
Good idea.

zmb3 committed:
Add TDP message 10 (MFA)

Per-session MFA works just like it does for SSH sessions, but rather
than wrapping the MFA JSON challenge in a protobuf (like SSH), we
package it in a TDP message.

zmb3 committed:
Create MFA codec interface

zmb3 committed:
Fix build

zmb3 committed:
Reduce some JSON decoding duplication

zmb3 committed:
General cleanup
